packages/vertexai/src/methods/chrome-adapter.ts

```diff
@@ -15,33 +15,146 @@
  * limitations under the License.
  */
 
-import {
-  EnhancedGenerateContentResponse,
-  GenerateContentRequest,
-  InferenceMode
-} from '../types';
+import { isChrome } from '@firebase/util';
+import { Content, GenerateContentRequest, InferenceMode, Role } from '../types';
 
 /**
  * Defines an inference "backend" that uses Chrome's on-device model,
  * and encapsulates logic for detecting when on-device is possible.
  */
 export class ChromeAdapter {
+  private isDownloading = false;
+  private downloadPromise: Promise<AILanguageModel | void> | undefined;
+  private oldSession: AILanguageModel | undefined;
   constructor(
     private aiProvider?: AI,
+    // TODO: mode can be required now.
     private mode?: InferenceMode,
     private onDeviceParams?: AILanguageModelCreateOptionsWithSystemPrompt
   ) {}
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  /**
+   * Convenience method to check if a given request can be made on-device.
+   * Encapsulates a few concerns: 1) the mode, 2) API existence, 3) prompt formatting, and
+   * 4) model availability, including triggering download if necessary.
+   * Pros: caller needn't be concerned with details of on-device availability. Cons: this method
+   * spans a few concerns and splits request validation from usage. If instance variables weren't
+   * already part of the API, we could consider a better separation of concerns.
+   */
   async isAvailable(request: GenerateContentRequest): Promise<boolean> {
-    return false;
+    // Returns false if we should only use in-cloud inference.
+    if (this.mode === 'only_in_cloud') {
+      return false;
+    }
+    // Returns false if the on-device inference API is undefined.
+    const isLanguageModelAvailable =
+      isChrome() && this.aiProvider && this.aiProvider.languageModel;
```

Comment:
Could we solely rely on feature detection of `window.ai.languageModel` here? Or do non-Chrome browsers offer `window.ai.languageModel`, but are incompatible with our implementation?

Reply:
> Or do non-Chrome browsers offer window.ai.languageModel, but are incompatible with our implementation?

That's exactly the concern 👍 

The Prompt API isn't standard yet. It's only available in Chrome. For comparison, Firefox is approaching the problem differently. The concern is if a non-Chrome browser explores AI using the same API as Chrome, but implements it in an unexpected way.

Once the API is standard, we can rely on feature detection.

Reply:
Copied this feedback into our burndown.